students/FlorentinPopescu/07_Mailroom_Part1_L3_FP.py

```diff
@@ -0,0 +1,162 @@
+    letter(x, amount)
+    return None
+
+#-----------------------------------------------
+def name_donation_letter(table):
+    while True:
+        x = input("\nEnter: donor's name; list; quit:>>\n")
+        if x == 'quit':
+            return (table)
+        elif x == 'list':
+            print ('\nExisting donors:\n', donor_list(table))
+        elif x not in donor_list(table) and x not in ['list','quit']:
+            # gardian - reject names containing special characters, pure numbers, or NaNs. 
+            if not check_name(x) or x.isdigit() or x == " ":
+                print("We are sorry, but '{}' does not seem to be a not a valid name. Please re-enter;".format(x))
+            else:
+                print ("\nAdd new donor: {}".format(x))
+                try:
+                    amount = input("Welcome! Please enter your generous donation:\n")
+                    amount = float(amount)
+                    #gardian - when donation is positive, add the donor to the list along with donation and print a thank-you note. 
+                    if amount > 0:
+                        add_donor(table, x)
+                        addDonation_sendLetter(table, x, amount)
+                    else: print ("We accept only positive donations. Please enter a donation to become one of our donors.")
+                except:
+                    print('\nWrong amount. Try again','\n')
+        elif x in donor_list(table) and x not in ['list','quit']:
+            try:
+                amount = input("Welcome back {}! Please enter your new donation:\n".format(x))
+                amount = float(amount)
+                if amount > 0:
+                    addDonation_sendLetter(table, x, amount)
+                #gardian - when an old donor tries to withdraw money inform him his money was spent.
+                elif amount < 0:
+                    print ('Money you donated to charity in the past was spent for noble causes, we cannot return.\n')   
+                else: 
+                    print ("We accept only non-zero donations.")
+            except:
+                print("\nWrong amount. PLease re-enter:\n")
+
+#-----------------------------------------------
+def report(table):
+    print("\n")
+    print("-"*77)
+    print ("".join([f"{'Donor Name':<20}", "|", f"{'Total Given |':>19}", f"{'Num Gifts |':>18}", f"{'Average Gift':>18}"]))
+    print("-"*77)
+    for item in table:
+        print(f"{item[0]:<19} | ${round(item[1],2):>15,.2f} | {item[2]:>15d} | ${round(item[3],2):>15,.2f}")
+    print("-"*77)
+    print("\n")
+
+#-----------------------------------------------
+def main(table, select):
+    while True:
+        for idx, choice in enumerate(select):
+             print(idx + 1, choice + '\n')
+        x = input('\nSelect one of the above: 1, 2, or 3:>>\n')
+        try:
+            x = int(x)
+            if x == 1: name_donation_letter(table)
+            elif x == 2: report(statistics(table))

```

Comment:
this will matter more in next version of mailroom but you'll want to have higher level functions from user selection like `report()` and `send_thank_you()`